### PR TITLE
[opencv.js doc] add doc for optimization and perf test

### LINKED ISSUE #279: [Doc] optimizations of OpenCV.js
The doc would include steps:

- [x] how to enable the multi-threads build
- [x] how to enable the SIMD build
- [x] how to test performance

### PR DESCRIPTION
**Pull request by Wenzhao-Xiang**
fix #279 

### This pullrequest changes
Add doc for how to enable opencv.js optimization and run perf test.

@huningxin @terfendail  PTAL, Thanks!

### CODE REVIEW

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 1]
```diff
@@ -103,3 +103,76 @@ Building OpenCV.js from Source
...
+    @code{.bash}
+    python ./platforms/js/build_js.py build_js --build_wasm --simd --build_wasm_intrin_test
+    @endcode
+
+    For wasm intrinsics tests, you can use the following function to test all the cases:
+    @code{.js}
+    cv.test_hal_intrin_all()
+    @endcode
+
+    And the failed cases will be print on the console.
```
Comment (huningxin) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown]: And the failed cases will be logged in the JavaScript debug console.

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 2]
```diff
@@ -103,3 +103,76 @@ Building OpenCV.js from Source
...
+    python ./platforms/js/build_js.py build_js --build_wasm --threads
+    @endcode
+
+    The default threads number is the logic core number of your device. You can use `cv.parallel_pthreads_set_threads_num(number)` to set threads number by yourself and use `cv.parallel_pthreads_get_threads_num()` to get the current threads number.
+
+    @note
+    You should build wasm version of `opencv.js` if you want to enable this optimization. 
+
+-#  [optional] To build `opencv.js` with wasm simd optimization, append `--simd` option.
+    To build wasm intrinsics tests, append `--build_wasm_intrin_test` option.
```
Comment (huningxin) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown]: Move this sentence after following example as it is only useful for development flow.

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 3]
```diff
@@ -103,3 +103,76 @@ Building OpenCV.js from Source
...
+
+    For example:
+    @code{.bash}
+    python ./platforms/js/build_js.py build_js --build_wasm --simd --build_wasm_intrin_test
```
Comment (huningxin) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown:144]: remove `--build_wasm_intrin_test`

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 4]
```diff
@@ -103,3 +103,76 @@ Building OpenCV.js from Source
...
+    @code{.bash}
+    python ./platforms/js/build_js.py build_js --build_wasm --threads
+    @endcode
+
+    The default threads number is the logic core number of your device. You can use `cv.parallel_pthreads_set_threads_num(number)` to set threads number by yourself and use `cv.parallel_pthreads_get_threads_num()` to get the current threads number.
+
+    @note
+    You should build wasm version of `opencv.js` if you want to enable this optimization. 
+
+-#  [optional] To build `opencv.js` with wasm simd optimization, append `--simd` option.
```
Comment (huningxin) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown:119]: Need to mention the simd optimization is experimental as wasm simd is still in development.

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 5]
```diff
@@ -103,3 +103,76 @@ Building OpenCV.js from Source
...
+    cv.test_hal_intrin_uint32()
+    cv.test_hal_intrin_int32()
+    cv.test_hal_intrin_uint64()
+    cv.test_hal_intrin_int64()
+    cv.test_hal_intrin_float32()
+    cv.test_hal_intrin_float64()
+    @endcode
+
+    @note
+    Now only emscripten LLVM upstream backend support wasm simd, so you need to setup upstream backend environment with the following command first:
```
Comment (huningxin) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown]: "support" -> "supports", it's better to have a link to https://emscripten.org/docs/porting/simd.html

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 6]
```diff
@@ -103,3 +103,76 @@ Building OpenCV.js from Source
...
+
+    For example:
+    @code{.bash}
+    python ./platforms/js/build_js.py build_js --build_wasm --threads
+    @endcode
+
+    The default threads number is the logic core number of your device. You can use `cv.parallel_pthreads_set_threads_num(number)` to set threads number by yourself and use `cv.parallel_pthreads_get_threads_num()` to get the current threads number.
+
+    @note
+    You should build wasm version of `opencv.js` if you want to enable this optimization.
```
Comment (huningxin) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown]: need to mention the threads optimization only works in browser, not in node.js

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 7]
```diff
@@ -103,3 +103,76 @@ Building OpenCV.js from Source
...
+
+    There are some kernels now in the performance test like `cvtColor`, `resize` and `threshold`. For example, if you want to test `threshold`, please navigate the web browser to `http://localhost:8080/perf/perf_imgproc/perf_threshold.html`. You need to input the test parameter like `(1920x1080, CV_8UC1, THRESH_BINARY)`, and then click the `Run` button to run the case. And if you don't input the parameter, it will run all the cases of this kernel.
+
+    You can also run tests using Node.js.
+
+    For example, run `threshold` with parameter `(1920x1080, CV_8UC1, THRESH_BINARY)`:
+    @code{.sh}
+    cd bin/perf
+    npm install
+    node threshold.js --test_param_filter="(1920x1080, CV_8UC1, THRESH_BINARY))"
```
Comment (huningxin) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown]: one more ")" at the end.

Comment (huningxin) [reply doc/js_tutorials/js_setup/js_setup/js_setup.markdown]: Will the filename be "perf_threshold.js"?

**Review by huningxin — CHANGES_REQUESTED** [review]
Finish 1st round of review. Please take a look at comments.

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 8]
```diff
@@ -103,3 +103,84 @@ Building OpenCV.js from Source
...
+
+    For example:
+    @code{.bash}
+    python ./platforms/js/build_js.py build_js --build_wasm --threads
+    @endcode
+
+    The default threads number is the logic core number of your device. You can use `cv.parallel_pthreads_set_threads_num(number)` to set threads number by yourself and use `cv.parallel_pthreads_get_threads_num()` to get the current threads number.
+
+    @note
+    You should build wasm version of `opencv.js` if you want to enable this optimization. And the threads optimization only works in browser, not in node.js. You should enable the `WebAssembly threads support` feature first with your browser.
```
Comment (huningxin) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown]: Please provide a link that tells how to enable `WebAssembly threads support`, e.g. in chrome.

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 9]
```diff
@@ -103,3 +103,84 @@ Building OpenCV.js from Source
...
+    Now only emscripten LLVM upstream backend supports wasm simd, refering to https://emscripten.org/docs/porting/simd.html. So you need to setup upstream backend environment with the following command first: 
+    @code{.bash}
+    ./emsdk update
+    ./emsdk install latest-upstream
+    ./emsdk activate latest-upstream
+    source ./emsdk_env.sh
+    @endcode
+
+    @note
+    You should build wasm version of `opencv.js` if you want to enable this optimization. And you should enable the `WebAssembly SIMD support` feature first.
```
Comment (huningxin) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown]: Please state that the `WebAssembly SIMD support` is in web browser.

**Review by huningxin — COMMENTED** [review]
Finish 2nd review. Please take a look at the comments.

**Comment by Wenzhao-Xiang** [comment]
@huningxin updated it.

**Comment by huningxin** [comment]
Looks good. Merge it so you can work on #295

**Pull request merged**